docs/decisions/0002-embedded-schemas.md

```diff
@@ -0,0 +1,42 @@
+# Embed schemas for reference lists using a custom schema notation
+
+## Context and problem statement
+
+Reference data lists have a schema, which is documented as prose in the appropriate sections of the README.
+
+It would be *nice* to have a machine-readable schema that lists can be checked against to ensure we don't make mistakes (eg, mistyping a field name in a hardcoded list).
+
+It is *essential* to have a machine-readable schema for automatic translation of the lists into forms that require an up-front schema, such as the BigQuery integration being developed.
+
+## Considered options
+
+* Convert the raw data into JSON and use http://json-schema.org/
+* Write a minimal schema notation of our own
+
+## Decision outcome
+
+Using a more general existing schema format would give us a nice-to-have validation solution, but we'd still need custom work to map it into other schemas such as BigQuery's schema model. We could only really map a subset of a JSON schema into bigquery's more limited relational structure, so we'd be restricted to a subset of JSON schema anyway.
+
+We only really use a small set of types/structures in reference data lists, and intend to keep it that way so we can easily map them into other formats.
+
+Therefore, I've written a minimal schema notation.
+
+### Example
+
+This is the schema for `Dfe::ReferenceData::Qualifications::QUALIFICATIONS`:
+
+```ruby
+{
+  :id => :string,
+  :name => :string,
+  :level => :string,
+  :suggestion_synonyms => { kind: :array, element_schema: :string },
+  :match_synonyms => { kind: :array, element_schema: :string },
+  :degree => { kind: :optional, schema: :string },
+  :hint => { kind: :optional, schema: :string }
```

Comment:
Could you document what the different types of `kind` might be?

And is it worth separating out `optional` into a separate thing? Logically, you might have a field which either contains an array or is null (ie an optional array), but perhaps that should be discouraged in favour of always including an empty array? (cue philosophical discussion on whether `[]` == `nil`)

It's also not quite clear to me what the difference between `element_schema` and `schema` is?

Reply:
Different types of `kind`: That's in the docs here: https://github.com/DFE-Digital/dfe-reference-data/pull/56/files#diff-79f5d80ee02a931b2bf12fd018b6edeb447abd58e1fb85ae155ae932ec29ad9d

Yeah, I gave some thought to whether to support optional arrays or just to encourage people to use empty arrays to mean "nothing", and decided to go for the latter in the end - the difference in semantic expressiveness is that we remove the ability for `nil` and `[]` to mean different things, and I felt that anybody actually _using_ that ability was probably making a terrible mistake :laughing:, probably using them as two different sentinel values ("user opted out of answering this question" vs "user hasn't been asked the question yet" sort of thing) in ways that would just be confusing to readers and best handled with an adjacent "status" field or some such. So decided to discourage such behaviour.

As for `element_schema` vs `schema` - purely that the array fields need a schema for the _elements_ of the array (an array of strings isn't a string) while the optional fields are still a string or whatnot just now made optional (an optional string is still a string). So I decided to use a different name for the field to emphasise this difference.